File: app/src/main/java/com/cn29/aac/ui/common/BaseRecyclerViewAdapter.java
package com.cn29.aac.ui.common;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.recyclerview.widget.RecyclerView;


/**
 * Created by Charles Ng on 6/9/2017.
 */

public abstract class BaseRecyclerViewAdapter<T, B extends ViewDataBinding>
    extends RecyclerView.Adapter<BaseRecyclerViewAdapter<T, B>.MyViewHolder<B>> {

  private final OnItemClickListener<T> itemClickListener;

  public BaseRecyclerViewAdapter(OnItemClickListener<T> itemClickListener) {
    this.itemClickListener = itemClickListener;
  }

  public BaseRecyclerViewAdapter<T, B>.MyViewHolder<B> onCreateViewHolder(ViewGroup parent,
      int viewType) {
    LayoutInflater layoutInflater =
        LayoutInflater.from(parent.getContext());
    B binding = DataBindingUtil.inflate(
        layoutInflater, viewType, parent, false);
    return new BaseRecyclerViewAdapter<T, B>.MyViewHolder<>(binding);
  }

  @Override
  public void onBindViewHolder(BaseRecyclerViewAdapter<T, B>.MyViewHolder<B> holder,
      int position) {
    final T item = getItemForPosition(position);
    holder.itemView.setOnClickListener(v -> itemClickListener.onItemClick(item));
    bind(holder.binding, item);
  }

  @Override
  public int getItemViewType(int position) {
    return getLayoutIdForPosition(position);
  }

  protected abstract T getItemForPosition(int position);

  protected abstract int getLayoutIdForPosition(int position);

  protected abstract void bind(B binding, T item);

  public interface OnItemClickListener<T> {

    void onItemClick(T item);
  }

  public class MyViewHolder<Binding extends ViewDataBinding> extends RecyclerView.ViewHolder {

    private final Binding binding;

    public MyViewHolder(Binding binding) {
      super(binding.getRoot());
      this.binding = binding;
    }


    public void bind(Binding binding, T item) {
      binding.executePendingBindings();
    }
  }
}

File: app/src/main/java/com/cn29/aac/ui/common/BaseRecyclerViewAdapter.kt
package com.cn29.aac.ui.common

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.cn29.aac.ui.common.BaseRecyclerViewAdapter.MyViewHolder

/**
 * Created by Charles Ng on 6/9/2017.
 */
abstract class BaseRecyclerViewAdapter<T, B : ViewDataBinding>(private val itemClickListener: OnItemClickListener<T>) :
        RecyclerView.Adapter<MyViewHolder<T, B>>() {
    override fun onCreateViewHolder(parent: ViewGroup,
                                    viewType: Int): MyViewHolder<T, B> {
        val layoutInflater: LayoutInflater = LayoutInflater.from(parent.context)
        val binding: B = DataBindingUtil.inflate(
                layoutInflater, viewType, parent, false)
        return MyViewHolder(binding)
    }

    override fun onBindViewHolder(holder: MyViewHolder<T, B>,
                                  position: Int) {
        val item = getItemForPosition(position)
        holder.itemView.setOnClickListener { v: View? ->
            itemClickListener.onItemClick(
                    item)
        }
        bind(holder.binding, item)
    }

    override fun getItemViewType(position: Int): Int {
        return getLayoutIdForPosition(position)
    }

    protected abstract fun getItemForPosition(position: Int): T
    protected abstract fun getLayoutIdForPosition(position: Int): Int
    protected abstract fun bind(binding: B,
                                item: T)

    interface OnItemClickListener<T> {
        fun onItemClick(item: T)
    }

    class MyViewHolder<T, Binding : ViewDataBinding>(val binding: Binding) :
            ViewHolder(binding.root) {
        fun bind(binding: Binding,
                 item: T) {
            binding.executePendingBindings()
        }

    }

}